usr/iscsid.c

```diff
@@ -202,11 +202,19 @@ static int sync_session(__attribute__((unused))void *data,
 	iscsiadm_rsp_t rsp;
 	struct iscsi_transport *t;
 	int rc, retries = 0;
+	char session_state[SCSI_MAX_STATE_VALUE];
 
 	log_debug(7, "sync session [%d][%s,%s.%d][%s]", info->sid,
 		  info->targetname, info->persistent_address,
 		  info->port, info->iface.hwaddress);
 
+	iscsi_sysfs_get_session_state(session_state, info->sid);
+
+	if (!strcmp(session_state, "FREE")) {
+		log_error("Skip syncing a freed session%d", info->sid);
+		return 0;
```

Comment:
The FREE state doesn't always mean we are freeing the session during removal. It means the session is no longer valid. This could be because it is being removed or the replacement/recovery timeout fired (see session_recovery_timedout in scsi_transport_iscsi.c). For the latter, if we re-login we will re-use the session struct. So if iscsid was restarted for the recovery timed out case we would still want to try to relogin/sync when iscsid starts up again.

I think you need to check if there is also a connection in sysfs (/sys/class/iscsi_connection). If there is no connection, then we were in FREE because it's being removed.

The problem above is that we've looped over sysfs and gathered the session and connection data, so we would expect that we would only get to this point if there was a valid connection. However, iscsi_sysfs_get_sessioninfo_by_id looks like it has a bug where it assumes that if we have a session then there will always be a connection, so if those

sysfs_get_str(ISCSI_CONN_SUBSYS)

calls fail, we just ignore it more or less so we can end up running sync_session.

So  I think we need 3 things:

1. In initiator.c iscsi_sync_session we could add some code that can check if there is a connection and session. If there is only a session and it's in FREE, then do like you did above and just return because it's being freed in the kernel.

2. If there is a connection and it's state is DOWN, then something was removing the session/connection but iscsid failed when we were removing the connection. Don't sync. Just finish the removal. If the conn state is DOWN then stop conn with STOP_CONN_TERM was already called. So we just need to do destroy_conn then destroy_session.

3. Fix up iscsi_sysfs_get_sessioninfo_by_id so it handles the case where the connection has been removed but the session is still there. Since for the above case, we want iscsid to cleanup we can't just return error and ignore the session. I'm not sure what we want to do here. Maybe just make sure the address/port are set to invalid values so the caller can figure out what to do?